lib/eph/blobLease.ts

```diff
@@ -0,0 +1,308 @@
+   */
+  partitionId?: string;
+  /**
+   * @property {string} [leaseId] The unqiue identifier for the lease. In the BlobLease this
+   * property will be set after getting a lease from the Azure Blob storage.
+   */
+  leaseId?: string;
+  /**
+   * @property {boolean} isHeld Determines whether the lease is held. Default: false.
+   */
+  isHeld: boolean;
+  /**
+   * Acquires the lease.
+   * @param {*} options Any options that need to be passed to acquire the lease.
+   * @returns {Promise<Lease>} Promise<Lease>
+   */
+  acquire(options: any): Promise<Lease>;
+  /**
+   * Renews the lease.
+   * @param {*} options Any options that need to be passed to renew the lease.
+   * @returns {Promise<Lease>} Promise<Lease>
+   */
+  renew(options: any): Promise<Lease>;
+  /**
+   * Releases the lease.
+   * @param {*} options Any options that need to be passed to release the lease.
+   * @returns {Promise<Lease>} Promise<Lease>
+   */
+  release(options: any): Promise<Lease>;
+  /**
+   * Updates the content
+   * @param {*} options Any options that need to be passed to update the blob content.
+   * @returns {Promise<Lease>} Promise<Lease>
+   */
+  updateContent(text: string, options?: any): Promise<Lease>;
+  /**
+   * Gets the content.
+   * @param {*} [options] Any options that need to be passed to get the blob content.
+   * @returns {Promise<string>} Promise<string>
+   */
+  getContent(options?: any): Promise<string>;
+}
+
+export class BlobLease implements Lease {
+
+  static notHeldError: string = "Lease not held";
+  static _beginningOfTime: string = new Date(1990, 1, 1).toUTCString();
+
+  partitionId?: string;
+  leaseId?: string;
+  storageAccount: string;
+  blobService: BlobService;
+  containerName: string;
+  blob: string;
+  fullUri: string;
+  hostName: string;
+
+  private _isHeld: boolean = false;
+  private _containerAndBlobExist: boolean = false;
+
+  constructor(hostName: string, storageConnectionString: string, containerName: string, blob: string) {
+    this.hostName = hostName;
+    this.blobService = createBlobService(storageConnectionString);
+    this.storageAccount = (storageConnectionString.match("AccountName=([^;]*);") || [])[1];
+    this.containerName = containerName;
+    this.blob = blob;
+    this.fullUri = `https://${this.storageAccount}.blob.core.windows.net/${containerName}/${blob}`;
+    this._isHeld = false;
+    this._containerAndBlobExist = false;
+    debug("[%s] Full lease path: '%s'.", this.hostName, this.fullUri);
+  }
+
+  /**
+   * Ensures that the container and blob exist.
+   */
+  async ensureContainerAndBlobExist(): Promise<void> {
+    try {
+      if (!this._containerAndBlobExist) {
+        await this._ensureContainerExists();
+        await this._ensureBlobExists();
+        this._containerAndBlobExist = true;
+      }
+    } catch (err) {
+      const msg = `An error occurred while ensuring that the container and blob exists. ` +
+        `It is: \n${JSON.stringify(err)}`;
```

Comment:
errors are notoriously bad at `JSON.stringify`ing. for example: `console.log(JSON.stringify(new Error('something went wrong')))` prints '{}'

you're probably better off doing `err ? err.stack : JSON.stringify(err)` so that things which are Error instances get a good string representation